### ISSUE TITLE
Probabilities of a novel image belonging to a Class

### ISSUE THREAD

**Issue by iou84** [sicara/easy-few-shot-learning#69]
I created an example that might help you.

```

import torchvision.transforms as tt
import torch
from torchvision.datasets import ImageFolder
from easyfsl.methods import FewShotClassifier
from torch.utils.data import DataLoader

class FewShotPredictor :
    """

        This class aims to implement a predictor for a Few-shot classifier.

        The few shot classifiers need a support set that will be used for calculating the distance between the support set and the query image.

        To load the support we have used an ImageFolder Dataset, which needs to have the following structure:

        folder:
          |_ class_name_folder_1:
                 |_ image_1
                 |_  …
                 |_ image_n
          |_ class_name_folder_2:
                 |_ image_1
                 |_  …
                 |_ image_n

        The folder must contain the same number of images per class, being the total images (n_way * n_shot).

        There must be n_way folders with n_shot images per folder.

    """

    def __init__(self ,
                 classifier: FewShotClassifier,
                 device,
                 path_to_support_images,
                 n_way,
                 n_shot,
                 input_size=224):

        """
            :param classifier: created and loaded model
            :param device: device to be executed
            :param path_to_support_images: path to creating a support set
            :param n_way: number of classes
            :param n_shot: number of images on each class
            :param input_size: size of image

        """
        self.classifier = classifier
        self.device = device

        self.predict_transformation = tt.Compose([
            tt.Resize((input_size, input_size)),
            tt.ToTensor()
        ])

        self.test_ds = ImageFolder(path_to_support_images, self.predict_transformation)

        self.val_loader = DataLoader(
            self.test_ds,
            batch_size= (n_way*n_shot),
            num_workers=1,
            pin_memory=True
        )

        self.support_images, self.support_labels = next(iter(self.val_loader))



    def predict (self, tensor_normalized_image):
        """

        :param tensor_normalized_image:
        Example of normalized image:

            pil_img = PIL.Image.open(img_dir)

            torch_img = transforms.Compose([
                transforms.Resize((224, 224)),
                transforms.ToTensor()
            ])(pil_img)

            tensor_normalized_image = tt.Normalize([0.485, 0.456, 0.406], [0.229, 0.224, 0.225])(torch_img)[None]


        :return:

        Return

        predict = tensor with prediction (mean distance of query image and support set)
        torch_max [1] = predicted class index

        """

        with torch.no_grad():
           self.classifier.eval()
           self.classifier.to(self.device)
           self.classifier.process_support_set(self.support_images.to(self.device), self.support_labels.to(self.device))
           pre_predict = self.classifier(tensor_normalized_image.to(self.device))
           predict = pre_predict.detach().data
           torch_max = torch.max(predict,1)
           class_name = self.test_ds.classes[torch_max[1].item()]
           return predict, torch_max[1], class_name

``` 

#49

_Originally posted by @diego91964 in https://github.com/sicara/easy-few-shot-learning/issues/17#issuecomment-1157091822_
     
Good morning guys and many thanks for the awesome and very helpful code and the effort put to achieve that
I have a question regarding novel image class prediction:
Is there a way to calculate as in 'classical' classification the percentage/probability of a novel image to belonging to each class?
Do you believe a softmax maybe at the  'return predict, torch_max[1], class_name' at the return tensor would have a meaning?

Thanks in advance

**Comment by ebennequin** [comment]
Hi @iou84 , thank you for your kind words!

I believe that applying a softmax on the `predict` object here is indeed what you need to get a prediction that can be interpreted as a probability distribution. 

Although, depending on the method, it can be more or less exploitable. For instance, with Prototypical Networks and using euclidean distance, the scores before softmax are likely to be large negative numbers, which would cause the output of your softmax to be very flattened (like `[0.20, 0.21, 0.20, 0.19, 0.20]` for 5 classes for instance). In that case I guess you would need to play with the softmax temperature, but I'm unsure that you can find a temperature that will suit all few-shot tasks.

**Comment by iou84** [comment]
Thank you very much for your quick response and advise
I have tried the softmax on my setting with Prototypical and gives good results (dominamt class around .88 -0.99!?)
Also can I ask lets say a more philosophical question:
Is there a way to infer if an uknown image doesnt belong to any of the classes?? (e.g. extremely high distance values  from all prototypes??)

Thanks
________________________________

**Comment by ebennequin** [comment]
So glad you asked! I happen to be working on this exact topic. You can find a first draft of our work [here](https://arxiv.org/abs/2206.09236), and the code [here](https://github.com/ebennequin/few-shot-open-set).
